### PR TITLE
Fix handshake mode for blockwise.

### PR DESCRIPTION
**Pull request by boaks**
Signed-off-by: Achim Kraus <achim.kraus@bosch-si.com>

### CODE REVIEW

**Comment by sbernard31** [comment]
I have a quick look at this PR and I think it should work better.

This code is very specific to HANDSHAKE_MODE and this create a kind of "dependency" between HANDSHAKE_MODE and transparent block wise.

To avoid that I proposed to add a kind of type to endpoint context attributes which define if attribute should be used once or for each "children" requests (https://github.com/eclipse/californium/issues/1186#issuecomment-575246399). This concept could works for transparent blockwise but also maybe for CoAP retransmission.  
Maybe using a prefix or suffix like KEY_PREFIX_NONE_CRITICAL.
But maybe this is a bit overkill as for now we don't have so much attributes :thinking:.

Talking about  CoAP retransmission, I guess we maybe face the kind of issue ? (not sure at all)

**Comment by boaks** [comment]
>But maybe this is a bit overkill

No, that may/will be the plan for a 3.0.

Changing the value of final String may cause much more stress.
Changing the key value of the handshake-mode using new prefixes, which makes that approach more generic breaks the API logically in a way, which gets hard to detect.
I'm not sure, if this fix is required for a 2.0.1, therefore I tried to change it in a "save way".

**Comment by sbernard31** [comment]
> I'm not sure, if this fix is required for a 2.0.1, therefore I tried to change it in a "save way".

:+1:

**Comment by boaks** [comment]
;-(

requires more work ... notifies from other addresses will require to use not the endpoint context of the original request, though that's destination is not longer valid.

 ;-).

**Comment by boaks** [comment]
After some investigation and tests, I consider to use the response's endpoint context and preserve the handshake-mode "none", is the most proper solution.

File: californium-core/src/main/java/org/eclipse/californium/core/network/stack/BlockwiseStatus.java [hunk 1]
```diff
@@ -231,6 +235,35 @@ final synchronized int getBlockCount() {
...
+			// considering notifies with address changes,
+			// use the response's endpoint-context to compensate that
+			EndpointContext context = exchange.getRequest().getDestinationContext();
+			String mode = context.get(DtlsEndpointContext.KEY_HANDSHAKE_MODE);
+			if (mode != null && mode.equals(DtlsEndpointContext.HANDSHAKE_MODE_NONE)) {
+				// restore handshake-mode "none"
+				blockContext = MapBasedEndpointContext.addEntries(blockContext, DtlsEndpointContext.KEY_HANDSHAKE_MODE,
+						DtlsEndpointContext.HANDSHAKE_MODE_NONE);
+			}
+			followUpEndpointContext = blockContext;
```
Comment (sbernard31) [inline californium-core/src/main/java/org/eclipse/californium/core/network/stack/BlockwiseStatus.java]: clearly a details but why not ? 

```java
if (mode != null && mode.equals(DtlsEndpointContext.HANDSHAKE_MODE_NONE)) {
	// restore handshake-mode "none"
	followUpEndpointContext = MapBasedEndpointContext.addEntries(blockContext, DtlsEndpointContext.KEY_HANDSHAKE_MODE, DtlsEndpointContext.HANDSHAKE_MODE_NONE);
} else {
	followUpEndpointContext = blockContext;
}
```

Comment (boaks) [reply californium-core/src/main/java/org/eclipse/californium/core/network/stack/BlockwiseStatus.java]: If you prefer ...

Comment (sbernard31) [reply californium-core/src/main/java/org/eclipse/californium/core/network/stack/BlockwiseStatus.java]: Don't do it if you don't think, it's clearer.

Comment (boaks) [reply californium-core/src/main/java/org/eclipse/californium/core/network/stack/BlockwiseStatus.java]: Done ... over the time I get used to the other style, mainly to initialize "final" stuff. But I have no real preference.

**Pull request merged**